### PR TITLE
feat(sdk-crash-detection): Event data allowlist

### PR DESCRIPTION
**Pull request by philippsuess**
Specify an allowlist for the event data to ensure only keeping essential data doesn't contain PII.

Fixes GH-49935

This PR is based on https://github.com/getsentry/sentry/pull/49928.

### CODE REVIEW

File: tests/sentry/utils/sdk_crashes/test_event_stripper.py [hunk 1]
```diff
@@ -75,10 +73,61 @@ def test_strip_event_data_strips_context(self):
...
-        assert contexts.get("app") is None
-        assert contexts.get("os") is not None
-        assert contexts.get("device") is not None
+        assert len(contexts) == 2
+
+        os_context = contexts.get("os")
+        assert len(os_context) == 3
+        assert os_context.get("name") == "iOS"
+        assert os_context.get("version") == "16.3"
+        assert os_context.get("build") == "20D47"
```
Comment (jjbayer) [inline tests/sentry/utils/sdk_crashes/test_event_stripper.py:82]: This should work

```suggestion
        assert contexts.get("os") == {
          "name": "iOS",
          "version": "16.3",
          "build": "20D47",
        }
```

File: src/sentry/utils/sdk_crashes/event_stripper.py [hunk 2]
```diff
@@ -1,26 +1,64 @@
...
+            "family": Allow.SIMPLE_TYPE,
+            "model": Allow.SIMPLE_TYPE,
+            "arch": Allow.SIMPLE_TYPE,
+        },
+        "os": {
+            "name": Allow.SIMPLE_TYPE,
+            "version": Allow.SIMPLE_TYPE,
+            "build": Allow.SIMPLE_TYPE,
+        },
+    },
```
Comment (jjbayer) [inline src/sentry/utils/sdk_crashes/event_stripper.py:53]: Nice!

File: src/sentry/utils/sdk_crashes/event_stripper.py [hunk 3]
```diff
@@ -37,11 +75,37 @@ def strip_event_data(event: Event, sdk_crash_detector: SDKCrashDetector) -> Even
...
+                stripped_data[data_key] = data_value
+            elif allowed is Allow.ALL:
+                stripped_data[data_key] = data_value
+            else:
+                continue
+
+        elif isinstance(data_value, Mapping):
+            stripped_data[data_key] = _strip_event_data_with_allowlist(
+                data_value, allowlist_for_data
+            )
```
Comment (jjbayer) [inline src/sentry/utils/sdk_crashes/event_stripper.py:106]: Can this function also descend into lists? I imagine that would be necessary.

File: src/sentry/utils/sdk_crashes/event_stripper.py [hunk 4]
```diff
@@ -1,26 +1,64 @@
...
+    "type": Allow.SIMPLE_TYPE,
+    "datetime": Allow.SIMPLE_TYPE,
+    "timestamp": Allow.SIMPLE_TYPE,
+    "platform": Allow.SIMPLE_TYPE,
+    "sdk": {
+        "name": Allow.SIMPLE_TYPE,
+        "version": Allow.SIMPLE_TYPE,
+        "integrations": Allow.NEVER.with_explanation("Users can add their own integrations."),
+    },
+    "exception": Allow.ALL.with_explanation("We strip the exception data separately."),
```
Comment (jjbayer) [inline src/sentry/utils/sdk_crashes/event_stripper.py:40]: This still seems dangerous. There could be fields in `exception` that we don't want to keep, right?

**Review by jjbayer — APPROVED** [review]
Great improvement, but it seems to me that having an `Allow.ALL` variant still leaves room for sensitive or simply unnecessary data sneaking through.

File: tests/sentry/utils/sdk_crashes/test_event_stripper.py (hunk 1 above)
Comment (philipphofmann) [reply tests/sentry/utils/sdk_crashes/test_event_stripper.py:82]: Thanks, TIL 😃

File: src/sentry/utils/sdk_crashes/event_stripper.py (hunk 4 above)
Comment (philipphofmann) [reply src/sentry/utils/sdk_crashes/event_stripper.py:40]: Yes, indeed. I created an issue (https://github.com/getsentry/sentry/issues/50710) to solve this in a subsequent PR.

File: src/sentry/utils/sdk_crashes/event_stripper.py (hunk 2 above)
Comment (philipphofmann) [reply src/sentry/utils/sdk_crashes/event_stripper.py:53]: Thanks 😃

File: src/sentry/utils/sdk_crashes/event_stripper.py (hunk 3 above)
Comment (philipphofmann) [reply src/sentry/utils/sdk_crashes/event_stripper.py:106]: Good catch 🥇 . Not yet, but I will have to add this functionality for exceptions https://github.com/getsentry/sentry/issues/50710

**Pull request merged**